SignalServiceKit/src/Util/OWSTypingIndicators.swift

```diff
@@ -0,0 +1,333 @@
+        private let recipientId: String
+        private let deviceId: UInt
+        private var displayTypingTimer: Timer?
+        var isTyping = false {
+            didSet {
+                AssertIsOnMainThread()
+
+                let didChange = oldValue != isTyping
+                if didChange {
+                    Logger.debug("isTyping changed: \(oldValue) -> \(self.isTyping)")
+
+                    notify()
+                }
+            }
+        }
+
+        init(recipientId: String, deviceId: UInt) {
+            self.recipientId = recipientId
+            self.deviceId = deviceId
+        }
+
+        func didReceiveTypingStartedMessage() {
+            AssertIsOnMainThread()
+
+            // If the client receives a ACTION=TYPING message:
+            //
+            // Cancel the displayTyping timer for that (sender, device)
+            // Display the typing indicator for that (sender, device)
+            // Set the displayTyping timer for 15 seconds
+            displayTypingTimer?.invalidate()
+            displayTypingTimer = Timer.weakScheduledTimer(withTimeInterval: 15,
+                                                          target: self,
+                                                          selector: #selector(IncomingIndicators.displayTypingTimerDidFire),
+                                                          userInfo: nil,
+                                                          repeats: false)
+            isTyping = true
+        }
+
+        func didReceiveTypingStoppedMessage() {
+            AssertIsOnMainThread()
+
+            // If the client receives a ACTION=STOPPED message:
+            //
+            // Cancel the displayTyping timer for that (sender, device)
+            // Hide the typing indicator for that (sender, device)
+            displayTypingTimer?.invalidate()
+            displayTypingTimer = nil
+            isTyping = false
+        }
+
+        @objc
+        func displayTypingTimerDidFire() {
+            AssertIsOnMainThread()
+
+            // If the displayTyping indicator fires:
+            //
+            // Cancel the displayTyping timer for that (sender, device)
+            // Hide the typing indicator for that (sender, device)
+            displayTypingTimer?.invalidate()
+            displayTypingTimer = nil
+            isTyping = false
+        }
+
+        func didReceiveIncomingMessage() {
+            AssertIsOnMainThread()
+
+            // If the client receives a message:
+            //
+            // Cancel the displayTyping timer for that (sender, device)
+            // Hide the typing indicator for that (sender, device)
+            displayTypingTimer?.invalidate()
+            displayTypingTimer = nil
+            isTyping = false
+        }
+
+        private func notify() {
+            NotificationCenter.default.postNotificationNameAsync(OWSTypingIndicatorsImpl.typingIndicatorStateDidChange, object: recipientId)
```

Comment:
I'm pretty sure we should actually be notifying using the thread id, so this is likely to change.